fem/coefficient.hpp

```diff
@@ -671,43 +671,120 @@ class MatrixRestrictedCoefficient : public MatrixCoefficient
+    coefficients or a scalar and a scalar coefficient */
 class SumCoefficient : public Coefficient
 {
 private:
+   double aConst;
    Coefficient * a;
    Coefficient * b;
 
    double alpha;
    double beta;
 
 public:
+   // Result is _alpha * A + _beta * B
+   SumCoefficient(double A, Coefficient &B,
+                  double _alpha = 1.0, double _beta = 1.0)
+      : aConst(A), a(NULL), b(&B), alpha(_alpha), beta(_beta) { }
+
    // Result is _alpha * A + _beta * B
    SumCoefficient(Coefficient &A, Coefficient &B,
                   double _alpha = 1.0, double _beta = 1.0)
-      : a(&A), b(&B), alpha(_alpha), beta(_beta) { }
+      : aConst(0.0), a(&A), b(&B), alpha(_alpha), beta(_beta) { }
+
+   void SetAConst(double A) { a = NULL; aConst = A; }
+   double GetAConst() const { return aConst; }
+
+   void SetACoef(Coefficient &A) { a = &A; }
+   Coefficient * GetACoef() const { return a; }
+
+   void SetBCoef(Coefficient &B) { b = &B; }
+   Coefficient * GetBCoef() const { return b; }
+
+   void SetAlpha(double _alpha) { alpha = _alpha; }
+   double GetAlpha() const { return alpha; }
+
+   void SetBeta(double _beta) { beta = _beta; }
+   double GetBeta() const { return beta; }
 
    /// Evaluate the coefficient
    virtual double Eval(ElementTransformation &T,
                        const IntegrationPoint &ip)
-   { return alpha * a->Eval(T, ip) + beta * b->Eval(T, ip); }
+   {
+      return alpha * ((a == NULL ) ? aConst : a->Eval(T, ip) )
+             + beta * b->Eval(T, ip);
+   }
 };
 
-/// Scalar coefficient defined as the product of two scalar coefficients
+/** Scalar coefficient defined as the product of two scalar coefficients or
+    a scalar and a scalar coefficient. */
 class ProductCoefficient : public Coefficient
 {
 private:
+   double aConst;
    Coefficient * a;
    Coefficient * b;
 
 public:
+   ProductCoefficient(double A, Coefficient &B)
+      : aConst(A), a(NULL), b(&B) { }
    ProductCoefficient(Coefficient &A, Coefficient &B)
-      : a(&A), b(&B) { }
+      : aConst(0.0), a(&A), b(&B) { }
+
+   void SetAConst(double A) { a = NULL; aConst = A; }
+   double GetAConst() const { return aConst; }
+
+   void SetACoef(Coefficient &A) { a = &A; }
+   Coefficient * GetACoef() const { return a; }
+
+   void SetBCoef(Coefficient &B) { b = &B; }
+   Coefficient * GetBCoef() const { return b; }
 
    /// Evaluate the coefficient
    virtual double Eval(ElementTransformation &T,
                        const IntegrationPoint &ip)
-   { return a->Eval(T, ip) * b->Eval(T, ip); }
+   { return ((a == NULL ) ? aConst : a->Eval(T, ip) ) * b->Eval(T, ip); }
+};
+
+/** Scalar coefficient defined as the ratio of two scalars where one or both
+    scalars are scalar coefficients. */
+class RatioCoefficient : public Coefficient
+{
+private:
+   double aConst;
+   double bConst;
+   Coefficient * a;
+   Coefficient * b;
+
+public:
+   RatioCoefficient(double A, Coefficient &B)
```

Comment:
Could you add a short comment saying that the ratio is computed as A / B?

Reply:
I'd be happy to. Thanks!